jlm/llvm/opt/push.cpp

```diff
@@ -47,412 +49,388 @@ class NodeHoisting::Statistics final : public util::Statistics
+  {
+    const auto loopVar = thetaNode->MapPreLoopVar(output);
+    if (rvsdg::ThetaLoopVarIsInvariant(loopVar))
+    {
+      return computeTargetRegion(*loopVar.input->origin());
+    }
 
-      // add consumers to worklist
-      for (const auto & argument : arguments)
-      {
-        for (const auto & user : argument->Users())
-        {
-          auto tmp = rvsdg::TryGetOwnerNode<rvsdg::Node>(user);
-          if (tmp && !hasPredecessors(*tmp))
-            wl.push_back(tmp);
-        }
-      }
+    if (isInvariantMemoryStateLoopVar(loopVar))
+    {
+      return computeTargetRegion(*loopVar.input->origin());
     }
 
-    region->prune(false);
+    return *output.region();
   }
-}
 
-static bool
-is_theta_invariant(const rvsdg::Node * node, const std::unordered_set<rvsdg::Output *> & invariants)
-{
-  JLM_ASSERT(dynamic_cast<const rvsdg::ThetaNode *>(node->region()->node()));
-  JLM_ASSERT(!hasPredecessors(*node));
-
-  for (size_t n = 0; n < node->ninputs(); n++)
+  // Handle gamma outputs
+  if (const auto gammaNode = rvsdg::TryGetOwnerNode<rvsdg::GammaNode>(output))
   {
-    JLM_ASSERT(dynamic_cast<const rvsdg::RegionArgument *>(node->input(n)->origin()));
-    auto argument = static_cast<rvsdg::RegionArgument *>(node->input(n)->origin());
-    if (invariants.find(argument) == invariants.end())
-      return false;
+    return Context_->getTargetRegion(*gammaNode);
   }
 
-  return true;
-}
-
-void
-push_top(rvsdg::ThetaNode * theta)
-{
-  auto subregion = theta->subregion();
-
-  // push out all nullary nodes
-  for (auto & node : subregion->TopNodes())
+  // Handle theta outputs
+  if (const auto thetaNode = rvsdg::TryGetOwnerNode<rvsdg::ThetaNode>(output))
   {
-    if (!has_side_effects(&node))
-      copy_from_theta(&node);
+    return Context_->getTargetRegion(*thetaNode);
   }
 
-  /* collect loop invariant arguments */
-  std::unordered_set<rvsdg::Output *> invariants;
-  for (const auto & lv : theta->GetLoopVars())
+  // Handle simple node outputs
+  if (const auto node = rvsdg::TryGetOwnerNode<rvsdg::SimpleNode>(output))
   {
-    if (lv.post->origin() == lv.pre)
-      invariants.insert(lv.pre);
+    return Context_->getTargetRegion(*node);
   }
 
-  // initialize worklist
-  Worklist wl;
-  for (const auto & lv : theta->GetLoopVars())
+  throw std::logic_error("Unhandled output type!");
+}
+
+rvsdg::Region &
+NodeHoisting::computeTargetRegion(const rvsdg::Node & node) const
+{
+  if (node.ninputs() == 0)
   {
-    auto argument = lv.pre;
-    for (const auto & user : argument->Users())
+
+    auto [constantNode, constantOp] =
+        rvsdg::TryGetSimpleNodeAndOptionalOp<rvsdg::ControlConstantOperation>(node);
+    if (constantOp)
     {
-      auto tmp = rvsdg::TryGetOwnerNode<rvsdg::Node>(user);
-      if (tmp && !hasPredecessors(*tmp) && is_theta_invariant(tmp, invariants))
-        wl.push_back(tmp);
+      // FIXME: We currently need to leave control constants where they are even though there is
```

Comment:
@caleridas Is there still any reason (was there ever?) why we model the ControlType with as a state kind? I would rather like to convert it to a value kind to avoid these problems.